read_utils.py

```diff
@@ -1052,6 +1057,267 @@ def parser_bwamem_idxstats(parser=argparse.ArgumentParser()):
+
+            colors = list(plt.rcParams['axes.prop_cycle'].by_key()['color']) # get the colors for this style
+            segment_color = colors[segment_num%len(colors)] # pick a color, offset by the segment index
+            plt.fill_between(range(prior_domain_max, domain_max), position_depths, [0]*len(position_depths), linewidth=0, antialiased=True, color=segment_color)
+
+        plt.title(plot_title, fontsize=font_size*1.2)
+        plt.xlabel("bp", fontsize=font_size*1.1)
+        plt.ylabel("read depth", fontsize=font_size*1.1)
+
+        # to squash a backend renderer error on OSX related to tight layout
+        if plt.get_backend().lower() in ['agg', 'macosx']:
+            fig.set_tight_layout(True)
+        else:
+            fig.tight_layout()
+
+        plt.savefig(out_plot_file, format=plot_format, dpi=DPI) #, bbox_inches='tight')
+        log.info("Coverage plot saved to: " + out_plot_file)
+
+    os.unlink(bam_aligned)
+    os.unlink(bam_sorted)
+
+    if not out_summary:
+        os.unlink(coverage_tsv_file)
+    
+
+def parser_plot_coverage(parser=argparse.ArgumentParser()):
+    parser = parser_plot_coverage_common(parser)
+    util.cmd.common_args(parser, (('loglevel', None), ('version', None), ('tmp_dir', None)))
+    util.cmd.attach_main(parser, plot_coverage, split_args=True)
+    return parser
+
+__commands__.append(('plot_coverage', parser_plot_coverage))
+
+def align_and_plot_coverage(out_plot_file, plot_format, plot_style, plot_width, plot_height, plot_title, base_q_threshold, mapping_q_threshold, max_coverage_depth, read_length_threshold, out_summary,
+                            in_bam, ref_fasta, out_bam=None, sensitive=False, min_score_to_output=None
+                            ):
+    ''' 
+        Take reads, align to reference with BWA-MEM, and generate a coverage plot
+    '''
+    if out_bam is None:
+        bam_aligned = mkstempfname('.aligned.bam')
+    else:
+        bam_aligned = out_bam
+
+    ref_indexed = mkstempfname('.reference.fasta')
+    shutil.copyfile(ref_fasta, ref_indexed)
+
+    bwa = tools.bwa.Bwa()
+    samtools = tools.samtools.SamtoolsTool()
+
+    bwa.index(ref_indexed)
+
+    bwa_opts = []
+    if sensitive:
+        bwa_opts + "-k 12 -A 1 -B 1 -O 1 -E 1".split()
+
+    map_threshold = min_score_to_output or 30
+
+    bwa_opts + ["-T", str(map_threshold)]
+
+    aln_sam = mkstempfname('.sam')
+    aln_sam_filtered = mkstempfname('.filtered.sam')
+
+    bwa.mem(in_bam, ref_indexed, aln_sam, opts=bwa_opts)
+
+    # @haydenm says:
+    # For some reason (particularly when the --sensitive option is on), bwa
+    # doesn't listen to its '-T' flag and outputs alignments with score less
+    # than the '-T 30' threshold. So filter these:
+    os.system("grep \"^@\" " + aln_sam + " > " + aln_sam_filtered)
+    os.system("grep \"AS:i:\" " + aln_sam + " | awk -v threshold=" + str(map_threshold) + " '{split($14, subfield, \":\"); if(subfield[3]>=threshold) print $0}' >> " + aln_sam_filtered)
```

Comment:
Yeah, those lines can definitely be refactored. They were pulled from Hayden's known-working script. `samtools view` is definitely the answer.